auth-cas/config.php

```diff
@@ -61,15 +61,33 @@ function getOptions() {
         'label' => $__('Service label'),
         'configuration' => array('size'=>60, 'length'=>100),
         'hint' => $__('The text "Login with {label}" will appear on the login
-            button. By default this is "CAS".'))),
+          button. By default this is "CAS".'))),
       'cas-name-attribute-key' => new TextboxField(array(
         'label' => $__('Name attribute key'),
-        'configuration' => array('size'=>60, 'length'=>100))),
+        'configuration' => array('size'=>60, 'length'=>100),
+        'hint' => $__('This maps to the "name" form field.'))),
       'cas-email-attribute-key' => new TextboxField(array(
         'label' => $__('E-mail attribute key'),
-        'configuration' => array('size'=>60, 'length'=>100))),
+        'configuration' => array('size'=>60, 'length'=>100),
+        'hint' => $__('This maps to the "email" form field.'))),
+      'cas-custom-attributes' => new TextareaField(array(
+        'label' => $__('Custom attributes'),
+        'configuration' => array('cols'=>60, 'rows'=>5, 'length'=>1000,
+          'html'=>false, 'placeholder'=>
+            "sAMAccountName,adusername\nmobile,mobile"),
+        'hint' => $__('Extra attributes that are not required to set up an
+          account or sign in. Pass comma-delimited pairs of 2 in the form
+          "claim-attribute,form-field", one pair per line.'))),
+      'cas-username-form-field' => new TextboxField(array(
+        'label' => $__('Username form field'),
+        'configuration' => array('size'=>60, 'length'=>100),
```

Comment:
I don't think we should allow any value other than `user` or `email`. osTicket really only supports those two, and it's a bit silly in that it assumes any value that doesn't contain a `@` is a username. So if someone, for example, set this field to `name`, they'd actually be provisioning usernames based on the `name` which has all kinds of confusing implications. *

We can use a dropdown to constrain the value. We also might want to change the wording to be simply "Login by attribute".

\* In particular, I could imagine a scenario where if the attribute were one that the user controlled, they could change that attribute to match someone else's username and then login as them.

Reply:
One particular reason I see this being useful is if you're in a weird environment where you have an internal username (mapped to the LDAP username) and a "service" username that is used for other things mapped to another attribute... Probably a stretch of course.

While I do also find the wording less than optimal, the difference with how it's set up is that the username field _not_ mapped to CAS attribute, it's instead mapped to the information passed to osTicket's Contact Information form. This could be changed but it was the easiest way to set it up from my perspective.

One issue with limiting on our side would be that the CAS attribute "name" could be _anything_, as you can easily rename the attribute in the claim. In my opinion it should be up to the system administrator to make sure they don't do something stupid like using a user-changeable attribute as the osTicket username. A warning could be provided to alert people of this potentiality.

Reply:
What if we made it `user` vs `email`, with `email` as the default. By default we pull `user` via `cas->getUser()`, like you do now, but we'd also allow `user` to be remapped via the custom attributes.

I think the only additional change you'd need to make is not forcefully override `user` in `getUsername()`. That is, only set it if it's not already set.

Reply:
I like it. That will make the change more explicit and less prone to error.

Reply:
Should the custom field be `user` or `username`? Since osTicket internally uses `username` I do like that better.

Reply:
Works for me. I don't feel strongly either way.